### ISSUE TITLE
Release a pure Julia version

### ISSUE THREAD

**Issue by timholy** [tkluck/StatProfilerHTML.jl#16]
From https://discourse.julialang.org/t/ann-haml-jl-rubys-favourite-templating-language-comes-to-julia/29066

> replace StatProfilerHTML.jl’s perl backend by a pure Julia version.

If you're doing that, is there some reason ProfileView.jl isn't adequate? Would be nice to know why. It has a "community-maintained" (=unmaintained) SVG backend, so in theory it can output HTML.

**Comment by tkluck** [comment]
Thanks for asking! Last I checked, ProfileView didn't make a [fully linked document like this](https://www.infty.nl/StatProfilerHTML.jl/example-output/sort-jl-f12752f26483db5810d9-line.html#S-home-tkluck-src-julia-usr-bin-share-julia-base-sort-jl-partition-273). Has that changed?

**Comment by tkluck** [comment]
I'd summarize the difference as ProfileView being able to show a flame graph in many different backends. By comparison, StatProfilerHTML only supports an html backend, but it gives a richer document including the line-by-line source code and links up and down the stack ("step out" and "step into").

Does that answer your question? I'll close this issue if it does.

**Comment by tkluck** [comment]
@timholy closing the issue. Feel free to re-open (or message me on Slack) if you'd like to discuss how the tools are (/can be more?) complementary.

**Comment by timholy** [comment]
Sorry for my absence. Yes, the SVG output of ProfileView is very limited. The GTK version has all the same bells and whistles, though presented somewhat differently.

One option would be to have the "named" package be for a specific frontend; I could deprecate the HTML backend of ProfileView. Is there any analysis code of ProfileView that might be useful to you? Or have you already implemented everything you need?

**Comment by tkluck** [comment]
> The GTK version has all the same bells and whistles, though presented somewhat differently.

Wow! I had no idea. I'll check it out and see if it makes sense to merge some of these efforts.

**Comment by timholy** [comment]
Update, I'm in the process of splitting ProfileView into 4 packages:

- [LeftChildRightSiblingTrees](https://github.com/JuliaCollections/LeftChildRightSiblingTrees.jl) was used for internal data handling. The details aren't terribly special (any k-ary tree structure would do equally well), but I cleaned it up and made it a standalone package.
- [FlameGraphs](https://github.com/timholy/FlameGraphs.jl): handles the computational side of producing a flamegraph. 
- ProfileSVG and ProfileView will be visualization/export functionality (linked from the FlameGraphs README)

FlameGraphs is GUI-agnostic so if you do want to collaborate on the "analysis" side it might be a place to do it. The `dev` (but not `stable`) docs have just built and been hosted, so you can read more about it if you want. I've submitted a registration request https://github.com/JuliaRegistries/General/pull/7997.

**Comment by tkluck** [comment]
@timholy I just pushed a version to `master` that relies on `ProfileSVG.jl` for the svg part and on `HAML.jl` for the html rendering. Thanks for making the flame graph stuff available in separate packages!

It seems that `ProfileSVG`'s output hangs in Firefox even for small stack traces. Are you seeing the same?

**Comment by tkluck** [comment]
Closing this issue as version 1.1 is now fully Julia.

For the record, I ended up using `FlameGraphs.jl` but replaced `ProfileSVG` functionality by a custom-rolled svg template. That made it easier for me to fix the performance issues I observed in my own use case.

Now this is pure Julia we can also explore how much of this package can be merged into / made a backend of `ProfileView.jl`. Happy to have that conversation whenever you have time for the topic.
